# Pandas Tutorial

### Overview
Pandas is a Python library that is useful for loading, cleaning, and analyzing two dimensional (table-like) data. Data scientists often use it to import data from a particular format (.csv, .xlsx, .txt), explore it, and transform it to extract meaning.

Pandas can be used in a regular Python script; however, many data scientists find that using the integrated development environment called Jupyter Notebook, which allows chunks of code to be run and explored at any step in the data exploration process.

This tutorial will cover the basics of using pandas to import data and explore it.

### Install pandas
Install pandas in the same way that you would install any other Python package using the command line:
```
pip install pandas
```

### Import the pandas package
Pandas can be imported in the same way that any other package is imported to a Python environment. It is common practice, however, to give pandas the alias "pd". I recommend adhering to this practice.

In [1]:
import pandas as pd

### Import the data
In this example, I downloaded one of the most classic datasets in a .csv format, the Titanic survival dataset. The file contains information about each passenger on the famous Titanic cruiseliner, including their passenger class, sex, embark location, and if they survived or not.

To start working with this data, we will locate where the .csv file is saved on the machine relative to where our Python/Jupyter Notebook and use the pandas method "read_csv" to import it. Importantly, we will also save the imported data to a Python varaible called "df", which stands for *dataframe*.

In [2]:
df = pd.read_csv("data/titanic.csv")

A dataframe is the pandas equivalent of a table. In that sense, a dataframe is composed of rows and columns, where each row is an observation and each column is a different attribute of the observation.

Now that the data was imported and saved into a variable called "df", we can print out the dataframe to see the data.
###### Note: In a standard Python script you would call print(df), but in a Jupyter Notebook we can just call the variable df to view its contents

In [3]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

### Initial observations of the data
At first glance of our table, we can see a lot of information. First, note that all the columns of the table have names, except for the leftmost one. The leftmost column in a dataframe is called the *index*, and it is used to attach a unique identifier to each row. It is used quite often to access specific rows of the dataframe.

Also note that the lower left-hand corner specifies that this dataset contains 891 rows and 12 columns. However, not all of them are presented in the data seen above. Pandas, by default, prevents users from displaying too many rows/columns of data, but this setting can be overridden.

It's often useful when exploring data to see the shape of the dataframe, or, in other words, how many rows or columns it has. We can see it visually above but sometimes it's helpful to get it programmatically as well.

In [4]:
df.shape

(891, 12)

We can also get some useful summary information from our dataframe using the .info() method, which tells us about each of our columns and their datatypes. "int64" indicates an integer, "object" indicates a string, and "float64" indicates a floating point number (decimal).

Observe that there are 11 columns (excluding the index column) and that not each column has data in all of its rows. For example, the "Cabin" column only has 204 non-null entries, even though there are 891 rows.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


There's also a method called ".describe()" which we can use to get standard information about columns in the dataframe with numerical information, including the count, mean, standard deviation, min, max, and quartiles.

In [6]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

At many points during the data exploration process you will likely need to refer to the original dataframe. To avoid printing out all of the rows, use the .head() function, which prints out the first five rows, or the .tail() function, which prints out the last 5 rows.

In [7]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [8]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

These statistics give us a general idea about our data and it's structure. Now let's look at how to drill down and aggregate data across columns and rows.
### Selecting individual columns and rows
Many times, data scientists only need to grab a single part of data to explore. This can be an individual column or a chunk of columns. It can also be an individual row or many rows.

To get an individual column out of a pandas dataframe, add the column name in square brackets after the name of the dataframe as exemplified below.

In [9]:
df['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888     NaN
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

To get back several columns in a new dataframe, replace the single column name (as seen above) with a list of column names.

In [10]:
df[['Age', 'Sex']]

Age     Sex
0    22.0    male
1    38.0  female
2    26.0  female
3    35.0  female
4    35.0    male
..    ...     ...
886  27.0    male
887  19.0  female
888   NaN  female
889  26.0    male
890  32.0    male

[891 rows x 2 columns]

To get an individual row, use the .loc[] property an pass in an index label. In this case, the index labels are numbers 0-890, although they can take other values.

In [11]:
df.loc[1]

PassengerId                                                    2
Survived                                                       1
Pclass                                                         1
Name           Cumings, Mrs. John Bradley (Florence Briggs Th...
Sex                                                       female
Age                                                         38.0
SibSp                                                          1
Parch                                                          0
Ticket                                                  PC 17599
Fare                                                     71.2833
Cabin                                                        C85
Embarked                                                       C
Name: 1, dtype: object

You can also get many rows this way by passing in a list of index label.

In [12]:
df.loc[[1,4,600]]

PassengerId  Survived  Pclass  \
1              2         1       1   
4              5         0       3   
600          601         1       2   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
600  Jacobsohn, Mrs. Sidney Samuel (Amy Frances Chr...  female  24.0      2   

     Parch    Ticket     Fare Cabin Embarked  
1        0  PC 17599  71.2833   C85        C  
4        0    373450   8.0500   NaN        S  
600      1    243847  27.0000   NaN        S

A similar dataframe property is .iloc[], which gets an individual row based on location rather than index label. In this case, the index label and the index location are the same.

In [13]:
df.iloc[1]

PassengerId                                                    2
Survived                                                       1
Pclass                                                         1
Name           Cumings, Mrs. John Bradley (Florence Briggs Th...
Sex                                                       female
Age                                                         38.0
SibSp                                                          1
Parch                                                          0
Ticket                                                  PC 17599
Fare                                                     71.2833
Cabin                                                        C85
Embarked                                                       C
Name: 1, dtype: object

In [14]:
df.iloc[[1,4,600]]

PassengerId  Survived  Pclass  \
1              2         1       1   
4              5         0       3   
600          601         1       2   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
600  Jacobsohn, Mrs. Sidney Samuel (Amy Frances Chr...  female  24.0      2   

     Parch    Ticket     Fare Cabin Embarked  
1        0  PC 17599  71.2833   C85        C  
4        0    373450   8.0500   NaN        S  
600      1    243847  27.0000   NaN        S

To get an individual cell, you can specify an index label and column name using .loc[], or an index number and column number using iloc[].

In [15]:
df.loc[1,'Name']

'Cumings, Mrs. John Bradley (Florence Briggs Thayer)'

In [16]:
df.iloc[600, 4]

'female'

### Aggregating and subsetting data
Aggregating data involves putting many data entries together to generate some information. Aggregation is a common practice in database operations and is easy in pandas, too.

There are some methods that make it easy to aggregate general information about columns in pandas. For example, the sum() method lets us quickly calculate the sum of a column with numeric data.

In [17]:
df['Fare'].sum()

28693.9493

There are other methods as well.

In [18]:
df['Name'].count()

891

In [19]:
df['Age'].mean()

29.69911764705882

Oftentimes, however, we want to aggregate information based on different groups of the dataframe. Just as you might group the data and run aggregate operations in a database query, you can also do it in pandas.

First, we will call the .groupby() method on the dataframe. This returns what is called a DataFrameGroupBy object, which we will then call the .agg() method on to run one or more aggregate operations.

In [20]:
df.groupby("Sex").agg({"Survived": ['mean', 'count']})

Survived      
            mean count
Sex                   
female  0.742038   314
male    0.188908   577

We can also subset the data in order to aggregate across specific groups. In this example, I will create a new dataframe called pclass_df that only includes passengers from class 3.

Subsetting occurs inside the dataframe by specifying one or more conditions to be met. Rows that meet the condition will be returned as a new dataframe. The syntax is as follows:

In [21]:
pclass_df = df[ df['Pclass'] == 3 ]
pclass_df

PassengerId  Survived  Pclass                                      Name  \
0              1         0       3                   Braund, Mr. Owen Harris   
2              3         1       3                    Heikkinen, Miss. Laina   
4              5         0       3                  Allen, Mr. William Henry   
5              6         0       3                          Moran, Mr. James   
7              8         0       3            Palsson, Master. Gosta Leonard   
..           ...       ...     ...                                       ...   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0      male  22.0      1      0         A/5 21171   7.2500   NaN        S  
2    female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4      male  35.0      0      0            373450   8.0500   NaN        S  
5      male   NaN      0      0            330877   8.4583   NaN        Q  
7      male   2.0      3      1            349909  21.0750   NaN        S  
..      ...   ...    ...    ...               ...      ...   ...      ...  
882  female  22.0      0      0              7552  10.5167   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
890    male  32.0      0      0            370376   7.7500   NaN        Q  

[491 rows x 12 columns]

As you can see, in this **new** dataframe, there are only 491 rows. Notice, however, that the index labels of each row stayed the same.

Now we can run a new aggregate query on the new dataframe.

In [22]:
pclass_df.groupby("Sex").agg({"Survived": ['mean', 'count']})

Survived      
            mean count
Sex                   
female  0.500000   144
male    0.135447   347

### Working with NULL and incorrect values 
Unclean data (data with null and incorrect values) is an inevitability. Living in an imperfect world with imperfect people and systems means that data cleaning will always be a necessity.

Null data (cells where data is absent) and dirty data (data with incorrect values) can be difficult to deal with, but luckily, pandas gives us some tools for dealing with them. For example, we can subselect at all the rows in our dataframe where Age is NAN (not a number) ie null.

In [23]:
df[ df['Age'].isna() ]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
5      male  NaN      0      0      330877   8.4583   NaN        Q  
17     male  NaN      0      0      244373  13.0000   NaN        S  
19   female  NaN      0      0        2649   7.2250   NaN        C  
26     male  NaN      0      0        2631   7.2250   NaN        C  
28   female  NaN      0      0      330959   7.8792   NaN        Q  
..      ...  ...    ...    ...         ...      ...   ...      ...  
859    male  NaN      0      0        2629   7.2292   NaN        C  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S  
868    male  NaN      0      0      345777   9.5000   NaN        S  
878    male  NaN      0      0      349217   7.8958   NaN        S  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S  

[177 rows x 12 columns]

Conversely, we can also subselect all the rows where age is not null.

In [24]:
df[ df['Age'].notnull() ]

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[714 rows x 12 columns]

Dropping rows with null values, however, isn't always the best solution to answering questions with data, especially when there are very few observations available. Data imputation involves inferring data based on other values in the data.

There are several ways to impute data, but let's say that instead of dropping the rows where age is null, we want to assign them the average age of everybody on the ship. We can apply functions to every row in a dataframe by using the .apply() function in combination with an anonymous (lambda) function. Note that we will also import the numpy library here to determine which rows are NAN.

In [25]:
# Import the numpy library
import numpy as np

# Get the average age
averageAge = df['Age'].mean()

# Make an anonymous function to return either age or average age, if age is null
assignMeanAge = lambda x: x if not np.isnan(x) else averageAge

# Apply the function to the column 'Age' of the dataframe. 
# Set the Age column of the dataframe equal to the new, imputed Age column.
df['Age'] = df['Age'].apply(assignMeanAge)

Now we can see that there are 891 non-null entries in the Age column

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### Exporting data
Sometimes data is imported into pandas and then used to feed machine learning models. Other times, data is imported into pandas to be modified and returned again as a .csv file.

Much in the same way that we imported our Titanic dataset, we can export a data frame to a .csv, .xlsx, and .txt file. We do so using the .to_csv() method of a dataframe.

The export method takes several parameters, but two are important. First, the location of the new file needs to be given. Second, the "index" parameter should be set (the default is True). Setting index to True (or not including it) will mean that the index labels of the rows will be included in the final csv file. This is important to note because many times, data files come without index labels and are expected to be returned in the same format that they were given.

For this reason, I will set the "index" parameter to False in the example.

In [27]:
df.to_csv("./data/newData.csv", index=False)

### Overview
In this tutorial, we covered how to load, clean, and analyze data with the pandas library in Python. We learned to create dataframes, explore them, subset them, and aggregate them to drill deeper into the data. We also learned how to deal with "dirty data" and export the final dataframe to a spreadsheet file for easy distribution with other business stakeholders.

Thanks for checking out this tutorial. If you have any questions or concerns, feel free to contact me at haydenthoopes@gmail.com.